[Merged by Bors] - Implement key cache to reduce keystore loading times for validator_client

blacktemplar opened the pull request:
## Issue Addressed

#1618 

## Proposed Changes

Adds an encrypted key cache that is loaded on validator_client startup. It stores the keypairs for all enabled keystores and uses as password the concatenation the passwords of all enabled keystores. This reduces the number of time intensive key derivitions for `N` validators from `N` to `1`. On changes the cache gets updated asynchronously to avoid blocking the main thread.

## Additional Info

If the cache contains the keypair of a keystore that is not in the validator_definitions.yml file during loading the cache cannot get decrypted. In this case all the keystores get decrypted and then the cache gets overwritten. To avoid that one can disable keystores in validator_definitions.yml and restart the client which will remove them from the cache, after that one can entirely remove the keystore (from the validator_definitions.yml and from the disk). 

Other solutions to the above "problem" might be:
* Add a CLI and/or API function for removing keystores which will update the cache (asynchronously).
* Add a CLI and/or API function that just updates the cache (asynchronously) after a modification of the `validator_definitions.yml` file.

Note that the cache file has a lock file which gets removed immediatly after the cache was used or updated.

paulhauner commented on validator_client/src/key_cache.rs:
```diff
@@ -0,0 +1,342 @@
+        &self.uuids
+    }
+
+    fn password(passwords: &[Vec<u8>]) -> Vec<u8> {
```
Zeroize

paulhauner commented on validator_client/src/key_cache.rs:
```diff
@@ -0,0 +1,342 @@
+
+    pub fn decrypt(
+        &mut self,
+        passwords: Vec<Vec<u8>>,
```
Zeroize

paulhauner commented on validator_client/src/key_cache.rs:
```diff
@@ -0,0 +1,342 @@
+    crypto: Option<Crypto>,
+    uuids: Vec<Uuid>,
+    #[serde(skip)]
+    pairs: HashMap<Uuid, Keypair>, //maps public keystore uuids to their corresponding Keypair
```
I would have thought it would be simpler to key this via `PublicKey` instead of `Uuid`. An EIP-2335 has a plain-text `pubkey` field you can access without the password. Perhaps I'm missing something.

blacktemplar replied:
The main idea here is to also consider the case that we have multiple keystores with the same PublicKey (but with different passwords and uuids, I assume if two keystores have the same uuid they are identical and also have the same password).

The goal is to keep the previous behavior (and basically don't make the change breaking), in this context especially this comment (https://github.com/sigp/lighthouse/blob/64e21c32eec426b418a947cdbbe460d35f64465e/validator_client/src/initialized_validators.rs#L486):

> A validator is considered "already known" and skipped if the public key is already known. I.e., if there are two different definitions with the same public key then the second will be ignored.

Currently I assume that two keystores with the same uuid are identical i.e. have the same secret and the same password ...

paulhauner commented on validator_client/src/initialized_validators.rs:
```diff
@@ -351,6 +396,84 @@ impl InitializedValidators {
+        }
+
+        //collect passwords
+        let mut passwords = Vec::new();
```
See later comment regarding Zeroize.

paulhauner commented on validator_client/src/initialized_validators.rs:
```diff
@@ -351,6 +396,84 @@ impl InitializedValidators {
+    ///
+    fn try_decrypt_key_cache(
+        &self,
+        cache: &mut KeyCache,
```
Taking the `cache` by ownership and returning it inside `Result::Ok` would avoid the need for the `bool` and make it more difficult to accidentally use a cache that failed decryption.

blacktemplar replied:
I wanted to distinguish between decryption failed because of a missing password which is a error not just for the cache but for the whole application (reading the single keystores won't help) and other reasons just concerning the cache in which case the cache can be ignored and the keystores read, but I assume this may be better designed by just adding new Error cases and handling them different.

paulhauner commented on validator_client/src/key_cache.rs:
```diff
@@ -0,0 +1,342 @@
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct KeyCache {
+    crypto: Option<Crypto>,
```
I'm not sure I understand why this is `Option`, is it to save some allocation when the cache isn't being used?

paulhauner reviewed (changes requested):
Looking good! I've said it before, I think this is quite an elegant solution, nice work.

I have a few comments/questions. Many of them are about using zeroize so we don't leave secret values floating around in memory.

I wonder if you've considered ignoring the cache if we have less than some `n` validators? Using the cache with 1 validator seems excessive. But if it's too complex then I'm sure the cache is fine.

CarlBeek commented:
I really like this way of handling this cache. It is elegant and IMO falls in a very reasonable position in the trade-off space. Nice job @blacktemplar!

blacktemplar commented on validator_client/src/key_cache.rs:
```diff
@@ -0,0 +1,342 @@
+    pub fn open<P: AsRef<Path>>(validators_dir: P) -> Result<Self, Error> {
+        let cache_path = validators_dir.as_ref().join(CACHE_FILENAME);
+        let file = OpenOptions::new()
+            .write(true)
```
You are right (and I fixed it now), I basically copied that from: https://github.com/sigp/lighthouse/blob/32338bcafa90e481fc841a94bb7c9016fdc34664/common/account_utils/src/validator_definitions.rs#L119-L124

So maybe we should also change it there :).

blacktemplar commented on validator_client/src/key_cache.rs:
```diff
@@ -0,0 +1,342 @@
+            .map(|(k, v)| (*k, v.sk.serialize().as_ref().into()))
+            .collect();
+
+        let raw = bincode::serialize(&secret_map).map_err(Error::UnableToSerializeKeyMap)?;
```
Added zeroize for `raw`. The question is: Is it ok to use `bincode::serialize` and `bincode::deserialize` if we don't know if this will copy the data at some point into a non-zeroize structure?

blacktemplar commented on validator_client/src/key_cache.rs:
```diff
@@ -0,0 +1,342 @@
+    state: State,
+}
+
+type SerializedKeyMap = HashMap<Uuid, Vec<u8>>;
```
Changed it to ZeroizeHash since this map holds secret keys. Note that I needed to add Serialize and Deserialize derives to ZeroizeHash to be able to serialize the map into a byte vector that we can encrypt and to deserialize the byte vector after decryption.

blacktemplar commented:
@paulhauner I addressed all your review comments and added zeroize for all relevenat variables storing secrets (note my comments on using bincode for serializing/deserializing the secrets map)

paulhauner commented on validator_client/src/initialized_validators.rs:
```diff
@@ -401,13 +543,50 @@ impl InitializedValidators {
+        let validators_dir = self.validators_dir.clone();
+        let log = self.log.clone();
+        if key_cache.is_modified() {
+            executor.spawn_blocking(
```
The `v0.3.0-staging` branch has this `update_validators` function, so we can await this `spawn_blocking` and avoid some race conditions.

paulhauner commented on validator_client/src/key_cache.rs:
```diff
@@ -0,0 +1,346 @@
+    }
+
+    pub fn add(&mut self, keypair: Keypair, uuid: &Uuid, password: PlainText) {
+        //do nothing in unencrypted state
```
I think "unencrypted" should be "not decrypted".

paulhauner reviewed (changes requested):
Looking good. I realized I missed something with `iv` reuse, I left a comment.

Also, in order to get this into the upcoming sec review we'll need to have this in the `v0.3.0-staging` branch. It looks like there's some conflicts when this targets `v0.3.0-staging`, so I won't be able to do a simple rebase to move this commit from master -> v0.3.0. Do you mind changing the base to `v0.3.0-staging`, please?

paulhauner reviewed (approved):
Nice! Once again, great solution.

As discussed offline there's still a bit of decryption happening on core tokio tasks, we should come back and move them into blocking tasks. A lot of this already existed before this PR.

bors r+
